common/comms/src/inbound_message_service/message_context.rs

```diff
@@ -0,0 +1,146 @@
+//  Copyright 2019 The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::inbound_message_service::{
+    comms_msg_handlers::determine_comms_msg_dispatch_type,
+    message_dispatcher::Dispatchable,
+};
+use serde_derive::{Deserialize, Serialize};
+use std::convert::TryFrom;
+use tari_comms::connection::message::{FrameSet, MessageEnvelopeHeader, MessageError};
+use tari_crypto::keys::PublicKey;
+
+/// The number of frames required to construct a MessageContext
+const FRAMES_PER_MESSAGE: usize = 5;
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct MessageContext<PubKey> {
+    pub connection_id: Vec<u8>,
+    pub source: Vec<u8>,
+    pub version: Vec<u8>,
+    pub node_identity: Option<PubKey>,
+    pub message_envelope_header: MessageEnvelopeHeader<PubKey>,
+    pub message_envelope_body: Vec<u8>,
+}
+
+impl<PubKey: PublicKey + 'static> MessageContext<PubKey> {
```

Comment:
Forcing PublicKey to be static seems wrong here. Doesn't this mean that the compiler will require every message's pubkey to live as long as the process? -- sounds like a defacto memory leak to me.

Reply:
https://www.reddit.com/r/rust/comments/8rhmj8/how_does_static_lifetime_work/
"The 'static lifetime implies that the object can outlive any lifetime. It doesn't necessarily mean it only gets dropped when the program terminates. It just means Rust doesn't have to enforce that this object goes out of scope before any other object does."

Pretty confusing - so the only options are perhaps: `'a` lifetime which could be elided and outlives the thread (same as 'static then?), a static lifetime (seems it doesn't always mean a value lives as long as the program) or a value which is owned by the thread (clone) - still don't really understand static lifetimes